lib/include/aws_iot_defender.h

```diff
@@ -0,0 +1,142 @@
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file aws_iot_defender.h
+ * @brief
+ */
+
+#ifndef _AWS_IOT_DEFENDER_H_
+#define _AWS_IOT_DEFENDER_H_
+
+/* Standard includes. */
+#include <stdint.h>
+
+/**
+ * This enum type is the ID of metrics. It can be specified in AwsIotDefender_SetMetrics function.
+ */
+typedef enum
+{
+    AWS_IOT_DEFENDER_METRIC_TCP_ESTABLISHED_TOTAL,
+} AwsIotDefenderMetric_t;
+
+/**
+ * Return codes of defender functions.
+ */
+typedef enum
+{
+    AWS_IOT_DEFENDER_SUCCESS = 0,
+    AWS_IOT_DEFENDER_ALREADY_STARTED,
+    AWS_IOT_DEFENDER_PERIOD_TOO_SHORT,
+    AWS_IOT_DEFENDER_OUT_OF_MEMORY,
+
+    AWS_IOT_DEFENDER_INTERNAL_FAILURE
+} AwsIotDefenderError_t;
+
+/**
+ * Root cause code of failure, used in callback
+ */
+typedef enum
+{
+    AWS_IOT_DEFENDER_FAILURE_NONE = 0,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_CREATION,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_CONNECTION,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_SUBSCRIPTION,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_PUBLISH,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_DISCONNECTION,
+    AWS_IOT_DEFENDER_FAILURE_MQTT_DELETION,
+    AWS_IOT_DEFENDER_FAILURE_METRICS_REJECTION
+} AwsIotDefenderFailureRootCause_t;
+
+/**
+ * Callback parameters.
+ */
+typedef struct AwsIotDefenderCallbackInfo
+{
+    const void * pPayload; /**< The received message if there is. */
+    size_t payloadLength;  /**< Length of the message. */
+    AwsIotDefenderFailureRootCause_t failureRootCause;
+} AwsIotDefenderCallbackInfo_t;
+
+/**
+ * Callback handle.
+ */
+typedef struct AwsIotDefenderCallback
+{
+    void * param1;
+    void ( * function )( void *,
+                         AwsIotDefenderCallbackInfo_t * const );
+} AwsIotDefenderCallback_t;
+
+/**
+ * Parameters of AwsIotDefender_Start function.
+ */
+typedef struct AwsIotDefenderStartInfo
+{
+    AwsIotDefenderCallback_t successCallback;
+    AwsIotDefenderCallback_t failureCallback;
+} AwsIotDefenderStartInfo_t;
+
+/**
+ * Helper macro for setting an array of metrics.
+ *
+ * metricsArray: must be a type of array. Do not pass pointer.
+ */
+#define AwsIotDefender_SetMetricsHelper( metricsArray ) \
+    AwsIotDefender_SetMetrics(                          \
+        metricsArray, sizeof( metricsArray ) / sizeof( AwsIotDefenderMetric_t ) )
+
+/**
+ * Set an array of metrics.
+ */
+AwsIotDefenderError_t AwsIotDefender_SetMetrics( const AwsIotDefenderMetric_t * const pMetricsArray,
+                                                 size_t metricsCount );
+
+/**
+ * Start the defender */
+AwsIotDefenderError_t AwsIotDefender_Start( AwsIotDefenderStartInfo_t * pStartInfo );
+
+/**
+ * Stop the defender
+ */
+AwsIotDefenderError_t AwsIotDefender_Stop( void );
+
+/**
+ * Set period in seconds.
+ */
+AwsIotDefenderError_t AwsIotDefender_SetPeriod( uint32_t periodSeconds );
```

Comment:
Should use `uint64_t` for time.

Reply:
You are saying uint32_t is too small?

Reply:
In the other libraries, time is in milliseconds, and for that uint32_t was too small.

I realize in Defender it must be in seconds... but just for consistency, I think we should make is uint64_t.

Reply:
ok